cellxgene_schema_cli/cellxgene_schema/validate.py

```diff
@@ -1282,6 +1292,251 @@ def _check_column_availability(self):
+            if "spot_diameter_fullres" not in uns_scalefactors:
+                self.errors.append(
+                    "uns['spatial'][library_id]['scalefactors'] must contain the key 'spot_diameter_fullres'."
+                )
+            # spot_diameter_fullres is specified: proceed with validation.
+            else:
+                self._validate_float(
+                    "uns['spatial'][library_id]['scalefactors']['spot_diameter_fullres']",
+                    uns_scalefactors["spot_diameter_fullres"],
+                )
+
+            # tissue_hires_scalef is required.
+            if "tissue_hires_scalef" not in uns_scalefactors:
+                self.errors.append(
+                    "uns['spatial'][library_id]['scalefactors'] must contain the key 'tissue_hires_scalef'."
+                )
+            # tissue_hires_scalef is specified: proceed with validation.
+            else:
+                self._validate_float(
+                    "uns['spatial'][library_id]['scalefactors']['tissue_hires_scalef']",
+                    uns_scalefactors["tissue_hires_scalef"],
+                )
+
+    def _has_no_extra_keys(self, dictionary: dict, allowed_keys: List[str]) -> bool:
+        """
+        Determine if the dictionary has only the given allowed keys. Keys can be missing (required
+        checks are executed separately) but no additional keys are allowed.
+
+        :param dict dictionary: the dictionary to check.
+        :param List[str] allowed_keys: the list of allowed keys.
+
+        :rtype bool
+        """
+        return set(dictionary.keys()).issubset(allowed_keys)
+
+    def _is_supported_spatial_assay(self):
+        """
+        Determine if the assay_ontology_term_id is either Visium (EFO:0010961) or Slide-seqV2 (EFO:0030062).
+
+        :return True if assay_ontology_term_id is Visium or Slide-seqV2, False otherwise.
+        :rtype bool
+        """
+        return self.adata.obs.get("assay_ontology_term_id").isin([ASSAY_VISIUM, ASSAY_SLIDE_SEQV2]).any()
+
+    def _is_valid_visium_image_shape(self, image: np.ndarray) -> bool:
+        """
+        Determine if the image has shape (,,3); image is expected to be a 3D numpy array
+        with the size of the last dimension being three.
+
+        :param np.ndarray image: the image to check the shape of.
+
+        :return True if image has shape (,,3), False otherwise.
+        :rtype bool
+        """
+        return len(image.shape) == 3 and image.shape[2] == 3
+
+    def _is_visium(self) -> bool:
+        """
+        Determine if the assay_ontology_term_id is Visium (EFO:0010961).
+
+        :return True if assay_ontology_term_id is Visium, False otherwise.
+        :rtype bool
+        """
+        assay_ontology_term_id = self.adata.obs.get("assay_ontology_term_id")
+        return assay_ontology_term_id is not None and (assay_ontology_term_id == ASSAY_VISIUM).any()
+
+    def _validate_float(self, name: str, value: float):
```

Comment:
see comments above; I think we can get rid of this.

If you disagree and would like to keep it, I think it'd make more sense to generalize into a `_validate_type` method instead and add `type` as a parameter. And generalize the docstring, which doesn't need to be specific to scalefactors